text/0113-virtual-script-container.md

```diff
@@ -0,0 +1,214 @@
+
+## Design
+
+A new element called `<virtual-script>` encapsulates the various mechanisms to evaluate the scripts in a virtual environment, which solves the problem when using synthetic shadow and subsequently, when using the native Shadow DOM. We can implement this solution in two phases:
+
+### Phase I: Synthetic Shadow
+
+Many LWC applications use the [synthetic-shadow polyfill](https://github.com/salesforce/lwc/tree/master/packages/%40lwc/synthetic-shadow/). This polyfill emulates the native Shadow DOM behavior while still allowing global styles to cascade into the shadow trees. Synthetic Shadow patches many of the DOM APIs that allow you to interact and traverse the DOM tree, (e.g.: `document.querySelector` and `document.querySelectorAll`), these APIs will prevent developers from penetrating the Shadow DOM boundaries defined by their LWC components. In this phase, we want to solve the LWC + Synthetic Shadow scenario.
+
+To ensure we can solve the problem while continuing to guarantee the benefits that are provided by the synthetic-shadow polyfill for all other components running on the main window, we are currently proposing the following characteristics for the virtual-script design:
+
+* Create a JS sandbox before the Synthetic Shadow Polyfill is evaluated.
+  * The sandbox is created using a same-domain iframe.
+
+* Use this JS sandbox to virtualize the DOM APIs in order to prevent these APIs to be subject to the Synthetic Shadow Polyfill patches:
+  * This can be done by caching references to the to-be-patched APIs inside the sandbox before the synthetic shadow is evaluated.
+
+* Evaluate one or more third party libraries inside this sandbox
+  * This allows the code to run in a non-patched environment, while the rest of the document is running synthetic shadow.
+
+* Package the above functionality in a web component which semantically matches the `<script>` tag API.
+
+* Provide an additional mechanism to remap globals when needed.
+    * The sandbox does NOT include any global name defined in the main window, whether it is defined before or after the creation of the sandbox. But the `<virtual-script>` can opt-in to share a global value between the sandbox and the main window, this must be done via the `extra-globals` argument passed to the `<virtual-script extra-globals="dataLayer">` element.
+
+Other considerations for Phase I:
+
+ * due to the explicit order of evaluation required for this to function, we plan to only support one sandbox that can be shared by multiple virtual scripts.
+ * budget is around 5kb (minify/gzip).
+ * performance penalties will be added to the virtual scripts as they run inside a membrane:
+   * evaluation budget: 30% slower than evaluating the same script in the main window.
+   * operation budget: 50% slower than executing the same task is the main window.
+   * 3rd party scripts are rarely part of the critical path for the application to function.
```

Comment:
How can you be so sure? Suggesting this is different from claiming it as fact.

Reply:
For the use cases we are targeting, this is a fact. i.e. B2C applications with third party integrations for tracking, analytics, or personalization. We aren't trying to load application framework code and that is not the use case for `virtual-script`

Reply:
maybe we can invert the statement... saying that with this tool, we are targeting non-critical path tools, and for that reason, we are going to force them to use `async` attribute, to fetch and evaluate when possible.